src/ValidatorRegistry.sol

```diff
@@ -0,0 +1,146 @@
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+// Interface
+import "./IValidatorRegistry.sol";
+
+/**
+ * @title ValidatorRegistry
+ * @dev Registry for managing validator contracts.
+ *      Implements UUPSUpgradeable for upgradability.
+ */
+contract ValidatorRegistry is
+    Initializable,
+    OwnableUpgradeable,
+    IValidatorRegistry,
+    UUPSUpgradeable
+{
+    // Mapping from validator address to validator name
+    mapping(address => string) private validatorNames;
+
+    // Events
+    event ValidatorRegistered(address indexed validator, string name);
+    event ValidatorRemoved(address indexed validator);
+
+    // Storage gap for future upgrades
+    uint256[50] private __gap;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @dev Initializes the ValidatorRegistry contract.
+     */
+    function initialize() public initializer {
+        __Ownable_init(msg.sender);
+        __UUPSUpgradeable_init();
+    }
+
+    /**
+     * @dev Authorizes the contract upgrade. Only the owner can upgrade.
+     * @param newImplementation Address of the new implementation contract.
+     */
+    function _authorizeUpgrade(address newImplementation)
+        internal
+        override
+        onlyOwner
+    {
+        // Authorization logic handled by onlyOwner modifier
+    }
+
+    /**
+     * @dev Registers a new validator.
+     * @param validator Address of the validator contract.
+     * @param name Name associated with the validator.
+     */
+    function registerValidator(address validator, string memory name)
+        public
+        onlyOwner
+    {
+        require(
+            validator != address(0),
+            "ValidatorRegistry: Invalid validator address"
+        );
+        require(
+            bytes(name).length > 0,
+            "ValidatorRegistry: Name cannot be empty"
+        );
+        require(
+            bytes(validatorNames[validator]).length == 0,
+            "ValidatorRegistry: Validator already registered"
+        );
+
+        validatorNames[validator] = name;
+        emit ValidatorRegistered(validator, name);
+    }
+
+    /**
+     * @dev Removes a validator from the registry.
+     * @param validator Address of the validator contract.
+     */
+    function removeValidator(address validator) public onlyOwner {
+        require(
+            bytes(validatorNames[validator]).length > 0,
+            "ValidatorRegistry: Validator not registered"
+        );
+
+        delete validatorNames[validator];
+        emit ValidatorRemoved(validator);
+    }
+
+    /**
+     * @dev Updates the name of a registered validator.
+     * @param validator Address of the validator contract.
+     * @param newName New name to associate with the validator.
+     */
+    function updateValidatorName(address validator, string memory newName)
```

Comment:
registerValidator and updateValidatorName can be the same function e.g. setValidatorName

Reply:
Isn't registering different than updating since updating something requires an existing registered item?

Reply:
Also this is assigning human readable names to the Validator Contract

Reply:
it only required this because of the restriction you add but technically speaking its the same onChain operation

Reply:
Ahh okay I see what you mean the restriction is for a level of redundancy especially if people mint without the UI or fork and build their own UI